Allow exiting map editor with Esc and show some non-rebindable keystrokes

CelticMinstrel opened the pull request:
This supercedes #467 - instead of hacking into the event handling to process SDL_QUIT, I have modified the terminate method to pass the key event into the SDL event queue (instead of pushing a quit event). This does have the weird side-effect that the OSX menu will lie about the key shortcut if someone decides to reassign Cmd+Q to something else, but I think that's probably not a major issue.

Also, unlike #467, this does not change the behaviour of clicking the close button on the window decorations. Since the main issue stems from how Cmd+Q is easy to hit by mistake, I don't think that's a major problem. Furthermore, Cmd+Q now does nothing while a dialog is up, which also differs from both #467 (where I attempted to have a confirmation in that event) and original behaviour (instant quit), but I think this actually makes sense - most apps can't be quit while a modal dialog is open.

There is one thing here that I'm not quite sure how to deal with - in SDLmain.m, I have constructed a pair of key events and pushed them into the SDL event queue, but I wonder if there's a better way to do this. In particular, the scancode probably shouldn't be -1 (aka 255) and it would be good if it could properly pass on which of the two command keys was pressed (though in practice the latter probably won't matter).

CelticMinstrel commented:
Okay, I went back to handling SDL_QUIT instead of modifying Cmd+Q to not produce one. (I'll rebase out that reverted commit later.) The current behaviour for me on OSX is as follows:


    Event   |Titlescreen |    Game    |   Editor   |   GUI1    |   GUI2
------------|------------|------------|------------|-----------|-----------
Escape      | Nothing    |Confirmation|Confirmation|Exit Dialog|Exit Dialog
Cmd+Q       | Terminate  |Confirmation|Confirmation| Nothing   | Nothing*
SIGINT      | Nothing    | Nothing    | Nothing    | Nothing   | Nothing
SIGTERM     | Terminate  |Confirmation|Confirmation| Nothing   | Nothing
Close Button|Confirmation|Confirmation| Nothing    | Nothing   | Nothing

*Pressing Cmd+Q in a GUI2 dialog appears to be read as a Q keystroke and inserts the letter into the active text field.  :| This seems like a separate bug that should perhaps be addressed.

The signals were tested via the kill command-line tool. Ctrl+C does not generate SIGINT on OSX in GUI applications, so it does nothing when pressed in all cases. The GUI1 dialog I tested was preferences; the GUI2 dialog I tested was the map editor's label dialog.

Personally, I think the only undesirable line there is SIGTERM - it should always just terminate, no confirmation. Provided the window managers on Linux don't implement the close button, Alt+F4, or whatever as sending SIGTERM, we can fix this by installing a signal handler for SIGTERM. We could add one for SIGINT too, if you want; I'm not sure if SDL generates an SDL_QUIT event from that on Linux, but it seems to be utterly ignoring it here on OSX.

CelticMinstrel commented:
I've forced Cmd+Q and Escape to be shown as bound hotkeys. I think it would be logical to do the same with Alt+F4, but I'm not sure how universal that shortcut is on Linux systems. Any insight on this?

gfgtdf commented:
> The current behaviour for me on OSX is as follows
You current as in in current master or do you mean at the current head or your PR ?
I dont know what sigterm or SIGINT is so i can only answer partly:

I think Cmd+Q/Alt+F4 shoudl always have ateh same effect as close button and having no effect on dialog is quite bad. Also on your sheet it looks like there is no option at all to close wesnoth during an open dialog.

CelticMinstrel commented:
That table refers to this pull request. Prior to the change, Escape did nothing in the editor and Cmd+Q unconditionally quit anywhere.

SIGINT and SIGTERM are UNIX things that by default cause a process to unconditionally terminate (a bit like using the Task Manager to shut it down on Windows, though on UNIX that effect would actually be SIGKILL). You're correct that the sheet indicates there's no way to close Wesnoth while a dialog is open (though I intend to make SIGTERM at least always quit unconditionally).

> I think Cmd+Q/Alt+F4 shoudl always have ateh same effect as close button

They do have the same effect right now, but it looks like I mixed up the table slightly in the close button row (it's fixed now).

> having no effect on dialog is quite bad.

How many programs do you know that you can quit while a modal dialog is up? I think it's reasonable to require the dialog to be dismissed before quitting. It's not even hard to do - just press Escape (possible more than once).

gfgtdf commented:
> How many programs do you know that you can quit while a modal dialog is up?
Wesnoth is different, dialogs are very common and also part of the game, specially since [message]s and maybe even right click menus are implemented as dialogs.
> require the dialog to be dismissed before quitting. It's not even hard to do 
Its not that easy, often (for examples is character dialogs but there also other places) the next dialog wil pop up after quiting a dialog.

> just press Escape 
If i remember correctly it is possible to deactivate quit on esc for some gui2 dialog explicitly.

CelticMinstrel commented:
You do have a point about messages. As I understand, those are a GUI2 dialog, so we could use the same method the titlescreen uses to make some dialogs, such as messages, objectives, etc, respond to quit keypresses. I'm not really sure which dialogs should support this, but I hope it's not too many; I don't think there's any reason to support it in any of the GUI1 dialogs, at least, and most dialogs that have buttons (especially OK/Cancel buttons or Yes/No buttons) probably don't need it. Right-click menus are GUI1, I believe, and it would make sense to respond to quit events while a menu is open, but I don't think it's a big deal not to, so (assuming I'm right) it can probably wait until the menus are GUI2.

If you're right that not all GUI2 dialogs can be exited via escape, I'm sure there's alternative ways to exit those dialogs before quitting, if it doesn't make sense to make them respond to the quit keypress.

gfgtdf commented:
> As I understand, those are a GUI2 dialog, so we could use the same method the titlescreen uses to make some dialogs, such as messages, objectives, etc, respond to quit keypresses. I'm not really sure which dialogs should support this, but I hope it's not too many

i don't see any advantages of having any dialogs not repsonding to quit and i don't think its easier to implement it only for some dialogs.

> right-click menus are GUI1, I believe, and it would make sense to respond to quit events while a menu is open, but I don't think it's a big deal not to, so (assuming I'm right) it can probably wait until the menus are GUI2.

I wouldn't count on right click menu beeing gui2 anytime soon so i really think it should react to quit during right click menus.

CelticMinstrel commented:
> i don't see any advantages of having any dialogs not repsonding to quit and i don't think its easier to implement it only for some dialogs.

I don't think there are advantages of not having dialogs in general (either GUI1 or GUI2) respond to quit events; however, when I attempted a general solution in the other branch, it got quite ugly. I don't think there are any disadvantages to dialogs in general _not_ responding to quit keystrokes, so I decided not to make them respond to the quit keystroke. It's odd in general for the program to be quittable while a modal dialog is up. For some dialogs (messages, objectives, victory, maybe a few others), I can see the need to make them handle quit keystrokes. I would want a complete list though before I go and implement that.

> I wouldn't count on right click menu beeing gui2 anytime soon so i really think it should react to quit during right click menus.

I can't see any reasonable way to make GUI1 in general react to quit keystrokes, but I may be able to code it in for menus only.

gfgtdf commented:
Hmm ok what was again teh reaon to mose the SDL_QUIT handler out of events.cpp and let everyone habdle it themselves?

CelticMinstrel commented:
Okay, I've cut out the part that actually deals with showing the quit confirmation (since gfgtdf's version of that was accepted instead), but there are still several other things in this PR that I think would be good.

* Escape key has the same effect in the editor as it does in the game.
* Quit to Desktop is shown in-game as well as in the editor.
* SIGTERM will skip the confirmation dialog

The Alt+F4 commit is untested since it would require testing on a non-windows machine. Additionally, the SIGTERM change is disabled on Windows machines, but I'm not certain this is actually necessary.

gfgtdf commented:
The Alt+F4 check checks whether the alt key is down, i think it would be better to also check that the ctrl key is not down, usually on windows, programms do not close when pressing alt+ctrl+F4

CelticMinstrel commented:
Did you test it? Does it trigger on Alt+Ctrl+F4? I don't think the code I have would do so, as it's comparing a bitmask with the equality operator... if anything, it's likely to have the problem that it won't work if for some reason you decide to hold both Alt keys and press F4. (Though I don't consider that a serious problem.)

Merely checking if the alt key is down would be `event.key.keysym.mods & KMOD_ALT`.

gfgtdf commented:
your right i read the code wonrgly. I see no reasons against merging then, but i also dont understand the SIGTERM stuff so i might not be teh best person to merge

aginor commented on src/events.hpp:
```diff
@@ -26,6 +26,7 @@
 #define DRAW_EVENT (SDL_USEREVENT + 3)
 #define CLOSE_WINDOW_EVENT (SDL_USEREVENT + 4)
 #define SHOW_HELPTIP_EVENT (SDL_USEREVENT + 5)
+#define TERMINATE_EVENT (SDL_USEREVENT + 6)
```
This uses up the second-to-last definable SDL_USEREVENT (giving it the value of 30, max is 31).

CelticMinstrel replied:
The other alternative I thought of would be to throw the exception from the signal handler, but I'm not sure if it's safe to do that.

aginor replied:
Can't you simply push an SDL_QUIT event onto the event queue?

aginor commented:
I dislike the way that the non-rebindable hotkeys are added. It introduces a special case in the signal handling code and the hotkey binding code. This leads to a conflict with PR #466, which would require an update to introduce the same changes.

I think a better solution would be to introduce a flag in the default game config that indicates whether a hotkey is rebindable or not. This flag can then be checked when rebinding hotkeys and deciding whether to allow it to succeed or not. The benefit of this approach would also be to have a future-proof way of adding more non-rebindable hotkeys.

gfgtdf commented:
@aginor 
I really think alt+F4 should be handled the same way as SDL_QUIT (i really think it's a bug in sdl that alt+F4 doesn't send SDL_QUIT by default) Adding this to the hotkey system is also nontrivial to do becasue it wouldn't work in dialogs (like [message] or right click menus) then.

CelticMinstrel commented:
@aginor It was already impossible to rebind Cmd+Q and Escape - all I added was an indication that the key exists but is not rebindable. Alt+F4 is a slightly different story; I basically agree with @gfgtdf that it should really raise an SDL_QUIT event, but since apparently it doesn't (at least on some systems), I added special handling to give it the same effect.

aginor commented:
@gfgtdf  you make a good point there.

However, I think that we also need to investigate the behaviour in SDL2. These changes should then only be included in the 1.12 branch if SDL2 does the right thing with ALT+F4/CMD+Q as the master branch will turn into 1.13.2 which will run SDL2 by default.

I did a quick test with SDL2 on Linux, and at least with my windowmanager it seems that ALT+F4 sends a SDL_QUIT.

gfgtdf commented:
> I did a quick test with SDL2 on Linux, and at least with my windowmanager it seems that ALT+F4 sends a SDL_QUIT.

In this case i think the simplest solution would be to but that alt+F4 in an #if sdl version check.

CelticMinstrel commented:
> I did a quick test with SDL2 on Linux, and at least with my windowmanager it seems that ALT+F4 sends a SDL_QUIT.

I was led to believe that this was already the case on SDL1.2 on Linux, but that on Windows an Alt+F4 did not send SDL_QUIT. I haven't tested Wesnoth on Windows though so I can't say if this is actually the case.

shikadilord commented:
The code looks fine to me, but since this touches code and behavior that's part of PR #466, I'd prefer to have @aginor review and merge it instead, as I said on IRC earlier.

aginor commented on src/wesnoth.cpp:
```diff
@@ -872,6 +876,15 @@ static std::vector<std::string> parse_commandline_arguments(std::string input)
+void wesnoth_terminate_handler(int sig);
+void wesnoth_terminate_handler(int sig) {
+	SDL_Event event;
+	event.type = TERMINATE_EVENT;
```
This could equally easy be SDL_QUIT instead of the user event, making the user event unnecessary.

CelticMinstrel replied:
If it was SDL_QUIT it would produce a confirmation dialog, which is what I'm trying to avoid. (In fact, it produces SDL_QUIT by default.)

aginor commented:
These changes conflicts with the newly created sdl2 branch that I'd like to merge them into (for that very reason). Would you mind changing the PR to be against that branch?
I'd also appreciate it if you could remove the user event and instead rely on the SDL_QUIT event that's already handled in the same way as the new user event.

CelticMinstrel commented:
Removing the user event renders the signal handler irrelevant and completely defeats the point of that commit. Unless you can think of a better idea to produce an instant quit upon receiving SIGTERM, I don't think removing the user event is an option.

aginor commented:
As we discussed on IRC, it sounds like either calling exit(0) or setting a global flag that is then detected in the event handling code is the right approach.

CelticMinstrel commented:
Note: Don't use the merge button now, since that will merge it into master rather than the sdl2 branch.

aginor commented on src/hotkey/hotkey_item.cpp:
```diff
@@ -24,537 +25,471 @@
+	}
+	
+	// These are hard-coded, non-rebindable hotkeys
+	if(id == "quit") {
```
I think something horrible happened here when you rebased. Is this snippet of code supposed to be in get_names() instead? There should also be a space between the if and the opening bracket to fit into the style in the rest of the file.

CelticMinstrel commented:
Fixed the mis-rebase. It still fails to build with a link error for missing _SDL_main. I googled around a bit but didn't find anything helpful yet.

aginor commented on src/wesnoth.cpp:
```diff
@@ -938,6 +948,15 @@ int main(int argc, char** argv)
+	struct sigaction terminate_handler;
+	terminate_handler.sa_handler = wesnoth_terminate_handler;
+	terminate_handler.sa_flags = 0;
+	terminate_handler.sa_mask = 0;
```
I'm getting a compile error on this line now, this is implemented as a struct in Linux so you cannot simply assign 0 to it.
`sigemptyset(&terminate_handler.sa_mask);` will do the right think on Linux and it looks like it's available on OSX too.

CelticMinstrel replied:
Weird, but I'll go ahead and make that change.

aginor commented:
https://github.com/aginor/wesnoth/commit/36e4f5610955b32ae61efd0825215f83fb69a3eb (untested) may fix the compilation issues you've hit around SDL_main. I don't think that should be treated as a blocker for sorting out the hotkey fixes for this PR (assuming we can test!).

This should be equally testable when compiled with SDL1.2 and SDL2 by the way, you will simply have inferior localised keyboard support on the sdl2 branch.

CelticMinstrel commented:
Thanks, that helped. The XCode build works now, and my cursory testing implies that the PR changes also work.

aginor commented:
Manually merged into branch sdl2.
